### PR TITLE
Add support for converting pyproject.toml-based Python3 packages.

### LINKED ISSUE #1780: fpm 1.12.0 Debian 11 ARM64: python to deb: easy_install executable not available, but required.
Environment:
```
# uname -a
Linux ng-pideb11 5.10.0-5-arm64 #1 SMP Debian 5.10.24-1 (2021-03-19) aarch64 GNU/Linux
```
```
# fpm --version
1.12.0
```
```
# pip show setuptools
Name: setuptools
Version: 56.0.0
```
```
# pip --version
pip 21.0.1 from /usr/local/lib/python3.9/dist-packages/pip (python 3.9)
```
```
# python3 --version
Python 3.9.2
```

Issue:

```
# fpm --verbose -s python -t deb --python-package-name-prefix python3 --python-bin=python3 argparse
Setting workdir {:workdir=>"/tmp", :level=>:info}
Trying to download {:package=>"argparse", :level=>:info}
Need executable 'easy_install' to convert python to deb {:level=>:error}
```

As stated on https://setuptools.readthedocs.io/en/latest/deprecated/easy_install.html , **easy_install** is deprecated. And removed from setuptools.
After some googling I wasn't able to get it from anywhere on Debian 11.

Are there any way to overcome this issue?
I.e. is there a way to use fpm to convert python packages to anything (.deb preferably)?

### LINKED ISSUE #1860: python: "Unexpected directory layout ..."
Forked this issue from #1780 

Relevant comments for this issue:
* https://github.com/jordansissel/fpm/issues/1780#issuecomment-966504491
* https://github.com/jordansissel/fpm/issues/1780#issuecomment-822975615

Summary: Converting python to deb is failing when using pip.

I'd like to figure out how to reproduce this, or at least dig a little deeper to help me understand where the problem in fpm might be.

Reported cases:
* `fpm --python-bin python3 --python-package-name-prefix python3 -s python -t deb --python-dis
able-dependency py-radix --iteration $CI_PIPELINE_ID aggregate6`
* `fpm -s python -t deb --python-package-name-prefix python3  --python-bin=python3 argparse`

### LINKED ISSUE #1873: Add suppport for pip's new `pyproject.toml` ?
For a while now it seems that `pip` has deprecated using `setup.py` in favor of a [different approach](https://pip.pypa.io/en/stable/reference/build-system/pyproject-toml/), using `pyproject.toml`. 

Would it be possible to support this approach for creating packages out of PyPI ? 

For example, the [structlog](https://github.com/hynek/structlog) project has completely discarded `setup.py` and trying to build a package for it fails:

```
$ fpm --verbose -s python -t deb --python-install-bin /usr/bin --python-bin=/usr/bin/python3 --python-package-name-prefix=python3 --python-pip=/usr/bin/pip3 --python-install-lib=/usr/lib/python3/dist-packages/ structlog
Setting workdir {:workdir=>"/tmp", :level=>:info}
Trying to download {:package=>"structlog", :level=>:info}
Looking in indexes: https://pypi.python.org/simple {:level=>:info}
Collecting structlog {:level=>:info}
  Using cached structlog-21.5.0.tar.gz (473 kB) {:level=>:info}
  Installing build dependencies: started {:level=>:info}
  Installing build dependencies: finished with status 'done' {:level=>:info}
  Getting requirements to build wheel: started {:level=>:info}
  Getting requirements to build wheel: finished with status 'done' {:level=>:info}
    Preparing wheel metadata: started {:level=>:info}
    Preparing wheel metadata: finished with status 'done' {:level=>:info}
  Saved /tmp/package-python-build-9d4a0e42c924109148bd3f697cf51ebdd039757d0336f4d6714e497a7639/structlog-21.5.0.tar.gz {:level=>:info}
Successfully downloaded structlog {:level=>:info}
Could not find 'setup.py' {:path=>"/tmp/package-python-build-9d4a0e42c924109148bd3f697cf51ebdd039757d0336f4d6714e497a7639/structlog/structlog/setup.py", :level=>:error}
Traceback (most recent call last):
	9: from /usr/local/bin/fpm:23:in `<main>'
	8: from /usr/local/bin/fpm:23:in `load'
	7: from /var/lib/gems/2.7.0/gems/fpm-1.14.1/bin/fpm:7:in `<top (required)>'
	6: from /var/lib/gems/2.7.0/gems/clamp-1.0.1/lib/clamp/command.rb:133:in `run'
	5: from /var/lib/gems/2.7.0/gems/fpm-1.14.1/lib/fpm/command.rb:574:in `run'
	4: from /var/lib/gems/2.7.0/gems/clamp-1.0.1/lib/clamp/command.rb:68:in `run'
	3: from /var/lib/gems/2.7.0/gems/fpm-1.14.1/lib/fpm/command.rb:351:in `execute'
	2: from /var/lib/gems/2.7.0/gems/fpm-1.14.1/lib/fpm/command.rb:351:in `each'
	1: from /var/lib/gems/2.7.0/gems/fpm-1.14.1/lib/fpm/command.rb:352:in `block in execute'
/var/lib/gems/2.7.0/gems/fpm-1.14.1/lib/fpm/package/python.rb:108:in `input': Unable to find python package; tried /tmp/package-python-build-9d4a0e42c924109148bd3f697cf51ebdd039757d0336f4d6714e497a7639/structlog/structlog/setup.py (RuntimeError)
```

Trying to create the package from an older version, when setup.py was still present, works:

```
$ fpm  -s python -t deb --python-install-bin /usr/bin --python-bin=/usr/bin/python3 --python-package-name-prefix=python3 --python-pip=/usr/bin/pip3 --python-install-lib=/usr/lib/python3/dist-packages/ structlog==20.2.0

Created package {:path=>"python3-structlog_20.2.0_all.deb"}
```

I'm testing on an up to date ubuntu focal, fpm 1.14.1, python3 3.8.2-0ubuntu2 , python3-pip 20.0.2-5ubuntu1.6

### PR DESCRIPTION
**Pull request by amdei**
TL;DR:
It works! (With some limitations at the moment though, as this is work in progress currently)
```
git clone ....
cd fpm
python3 -m pip install importlib_metadata
ruby bin/fpm --verbose --log debug --debug --debug-workspace -s python -t deb --python-install-lib=/usr/local/lib/python3.9/dist-packages/ --python-package-name-prefix python3 --python-bin=python3  typing_extensions
```

Limitations:

1. Does not support python2 in any way (obviously)

Temporary limitations (TODOs):

1. Do not support automatic dependencies extraction to the moment
2. Do not support license extraction to the moment
3. Checked only on Debian 11 amd64 (arm64 to go) so far
4. Not sure about applicability to 'native' packages
5. A lot of @todo's in code
6. Code quality is poor
7. Code-style is poor

I had no previous experience with ruby, lo any suggestions/comments/guidelines are welcome.

@ObjatieGroba - could you please check Python part of code, and give your suggestions on how to improve it?

Fixes #1780
Fixes #1873
Fixes #1860
Fixes magma/magma#13847

### CODE REVIEW

**Comment by amdei** [comment]
It seems to be working a little bit better now.
At least in terms of metadata retrieval.

Things to do:

1. Apply proper arch attribute to generated deb-package. At the moment all toml-based packages has `Architecture: all`;
2. Instruct wheel to not download dependencies for builded package wheel;
3. For some packages fpm now silently dies with zero exit code. Need to get it work again. It happens namely for latest `django`, `pyyaml`, `prettytable`, `requests`, `python-ldap`, `aiohttp`, `aiosignal`, `yarl` and probably lot of others;
4. Deal with packages dependencies. Not easy at the first glance:

For example for `psutil` python package dependencies looks like this:
```
REQ DIST: ['ipaddress ; (python_version < \"3.0\") and extra == \\'test\\'', 'mock ; (python_version < \"3.0\") and extra == \\'test\\'', 'enum34 ; (python_version <= \"3.4\") and extra == \\'test\\'', 'pywin32 ; (sys_platform == \"win32\") and extra == \\'test\\'', 'wmi ; (sys_platform == \"win32\") and extra == \\'test\\'']
```
Here is a `sys_platform`, `python_version`, `extra`, etc. - need to handle all of them.

Any suggestions/comments/guidelines are still welcome.
As well as package samples, where things go wrong.

**Comment by jordansissel** [comment]
Thanks for your work on this! I’ll take a look as soon as I can :)

**Comment by amdei** [comment]
In dry matter:

1. It is slow. Very slow. Significantly slower that setup.py-based packaging;
2. Wasn't able to made it with architecture - it looks like there is simply no way to get arch neither from `*.whl` file nor from `sdist` via any API. But it contains in `*.whl`. But I'm not brave enough to parse it in that way (yet). 
3. There is no alternatives found for setup.py's `--install-lib`, `--install-data`, `--install-scripts` and `build_scripts --executable`. So all fpm's `--python-install-*` options will not work (and currently silently ignored).
4. Staging with pip is broken. For example see here: https://github.com/pypa/pip/issues/8799 Or it is intended for something else.
5. Documentation for `pip install` is rather short: https://docs.python.org/3/installing/index.html#installing-index Significantly shorter than one would expect.

**Comment by amdei** [comment]
After a while, I hack the Wheel metadata to get information if package is pure or not.
https://peps.python.org/pep-0491/
As it seems to be impossible (at least yet) via pkginfo.Wheel API: https://bugs.launchpad.net/pkginfo/+bug/2015657

File: lib/fpm/package/pyfpm_wheel/get_metadata_wheel.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+            for operator, version in dep.specs:
+                deps.append("%s %s %s" % (dep.project_name,
+                        operator, version))
+        else:
+            deps.append(dep.project_name)
+
+        return deps
+
+    def get_home_url(self, project_urls):
+        res = dict([i.strip() for i in x.split(',')] for x in project_urls)
```
[the remark this answers is not in the document]
Comment (amdei) [reply lib/fpm/package/pyfpm_wheel/get_metadata_wheel.py:38]: It appears that different packages put home URL into different keys.
For example `Home` vs. `Homepage`.

So let's stick to a current approach, if there is no more efficient solutions...

**Comment by amdei** [comment]
It reached "works for me" stage.
Let's collect feedback before starting to polish things..

File: lib/fpm/package/pyfpm/get_metadata.py [hunk 2]
```diff
@@ -83,14 +83,17 @@ def run(self):
 
         if self.load_requirements_txt:
             requirement = open(self.requirements_txt).readlines()
+#            print('REQ-SETUP-PY-REQ-TXT:', requirement, file=sys.stderr)
```
Comment (ObjatieGroba) [inline lib/fpm/package/pyfpm/get_metadata.py:86]: unnecessary lines

File: lib/fpm/package/python.rb [hunk 3]
```diff
@@ -99,17 +103,28 @@ def input(package)
...
+      pyproject_toml = File.join(path_to_package, "pyproject.toml")
     else
       setup_py = path_to_package
+      pyproject_toml = path_to_package
     end
 
-    if !File.exist?(setup_py)
-      logger.error("Could not find 'setup.py'", :path => setup_py)
-      raise "Unable to find python package; tried #{setup_py}"
+    # @todo Swap it! - TOML priority is for debugging only!
```
Comment (ObjatieGroba) [inline lib/fpm/package/python.rb:112]: It is time to swap

File: lib/fpm/package/python.rb [hunk 4]
```diff
@@ -99,17 +103,28 @@ def input(package)
...
+    if File.exist?(pyproject_toml)
+      logger.debug("Do job with pyproject.toml")
+      wheel_path = build_py_wheel(setup_py)
+      load_package_info_wheel(setup_py, wheel_path)
+      # install_to_staging_toml(setup_py)
```
Comment (ObjatieGroba) [inline lib/fpm/package/python.rb:117]: unnecessary comment

File: lib/fpm/package/python.rb [hunk 5]
```diff
@@ -195,8 +210,185 @@ def download_if_necessary(package, version=nil)
...
+        "gmt = get_metadata_wheel.get_metadata_wheel('#{wheel_path}')",
+        "gmt.run('#{tmp}')"
+      ].join("\n")
+
+      get_metadata_cmd = "env PYTHONPATH=#{pylib}:$PYTHONPATH #{attributes[:python_bin]} " \
+        " -c " \
+        "#{Shellwords.escape(toml_metadata_code)}"
+
+      # @todo FIXME!
+      #      if attributes[:python_obey_requirements_txt?]
```
Comment (ObjatieGroba) [inline lib/fpm/package/python.rb:314]: Doesn't this break backward compatibility?

File: lib/fpm/package/python.rb [hunk 6]
```diff
@@ -322,9 +514,57 @@ def fix_name(name)
...
+
+    prefix = "/"
+    prefix = attributes[:prefix] unless attributes[:prefix].nil?
+
+    # Some setup's assume $PWD == current directory of pyproject.toml, so let's
+    # chdir first.
+    ::Dir.chdir(project_dir) do
+      flags = [ "--root", staging_path ]
+
+      # if !attributes[:python_install_lib].nil?
```
Comment (ObjatieGroba) [inline lib/fpm/package/python.rb:530]: Lot's of commented code. Is it necessary?

**Comment by JordanStopford** [comment]
Amy update on this? A lot of packages no longer ship setup.py files so if we need to use fpm then we're stuck on old versions

**Comment by cwegener** [comment]
I don't understand what this does? It still requires a legacy `setup.py` file to be generated by the python project that is supposed to be converted. However, pretty much all PEP-517 build backends disable the support for generating `setup.py` legacy files by default, so they will not be able to be used by `fpm`. (some backends don't even offer an option at all to generate legacy `setup.py` files)

And if you have a source python project that has the option to generate a legeacy `setup.py` enabled, then such a project can already be used as a source in `fpm` without this change.

I am a bit confused what this change is supposed to do.

What is the intention of this change?

**Comment by cwegener** [comment]
> > It still requires a legacy setup.py file
> 
> No, it doesn't.
> 
> There is an option to use `setup.py file`, if it presented. But if there is `pyproject.toml` file - it will be used instead.

Thanks for the clarification. I'll need to read through the patch again then. :grin:

**Comment by amdei** [comment]
Although in https://github.com/jordansissel/fpm/pull/1982 I made attempt to solve it via pip wheel, not via `build` module from PyPA, as suggested in https://github.com/jordansissel/fpm/issues/2040.

**Comment by cwegener** [comment]
> Although in this MR I made attempt to solve the issue via `pip wheel`, not via `build` module from PyPA, as suggested in #2040.

That's excellent. `pip wheel` will actually just invoke the Python `build` module which takes care of performing the right build process (legacy setup.py vs new PEP-517 build backends)

I will need to test out your PR and report the results.

Specifically, I will need to test with this PEP-517 project ... https://github.com/openai/openai-python

**Comment by gmabey** [comment]
Just an enthusiastic vote for more eyeballs on this merge request ... I would love :heart: to see the project catch up to this  shift in the python community.

**Comment by cwegener** [comment]
> I will need to test out your PR and report the results.

Quick feedback from my initial testing of this patch:

Testing with
- `openai` package from PyPI (which uses hatchling as the PEP 517 build backend)
- Testing on Ubuntu 22.04 with built-in Python3.10

My main issue that I'm looking into at the moment is the boot-strapping of the PEP 517 build environments for `pip wheel` (and to a lesser degree `pip download` ... which is a separate story).

I think the main concern so far is the new dependency on the `pkginfo` Python library. E.g. on Ubuntu 22.04, the version of `python3-pkginfo` is a bit older and does not yet have support for version 2.3 Metadata file formats.

But a lot of wheels will now write 2.3 Metadata file formats.

Therefore, injecting or vendoring an up-to-date version of `pkginfo` is important.

**Comment by amdei** [comment]
> My main issue that I'm looking into at the moment is the boot-strapping of the PEP 517 build environments for `pip wheel`

@cwegener could you please elaborate, what is your issue with ```openai``` package building?
Is it difficulty with determining build dependencies?
Or something else?

**Comment by cwegener** [comment]
> > My main issue that I'm looking into at the moment is the boot-strapping of the PEP 517 build environments for `pip wheel`
> 
> @cwegener could you please elaborate, what is your issue with `openai` package building? Is it difficulty with determining build dependencies? Or something else?

I think it's actually a different issue that is unrelated to the PEP-517 build support.

The issue is that **all** dependencies are being built from source due to a change introduced here: https://github.com/jordansissel/fpm/commit/77e0cf7a768bdd883409d58e0549691a7d1d762f

The `--no-build :all:` option forces all dependencies to be built from source, even though this is unnecessary.

The dependencies will be thrown away at the end. They don't end up in the released package that fpm creates. So there is really no benefit in building them from source.

Nevertheless, I still have not managed to build an `openai` package, but I think that might just be an issue on my side.

**Comment by amdei** [comment]
> Nevertheless, I still have not managed to build an `openai` package, but I think that might just be an issue on my side.

Just checked - built seamlessly.

Could you please try to build it with more verbose output?
Something like this:
```./bin/fpm  --verbose --log debug --debug --debug-workspace -s python -t deb --python-bin=python3 --python-package-name-prefix python3 --python-install-lib=/usr/local/lib/python3.9/dist-packages/ openai```

(you may want to adjust ```--python-install-lib``` option to reflect your python version)

**Comment by cwegener** [comment]
I had another look. And I think I figured out where the issue with this approach is.
`pip download` will perform unwanted build steps in order to generate metadata. Those unwanted build steps will fail quite often.

Example package: `jiter`

```
python3 -m pip download --no-clean --no-deps --no-binary :all: -d . jiter
```

This will try to immediately build a wheel from the `jiter` sdist, which will fail due to the build backend (`maturin`) not being present on my system.

This is a `pip` problem as per the following references.

References:
https://github.com/pypa/pip/issues/7995
https://github.com/pypa/pip/issues/1884

**Comment by amdei** [comment]
> And I think I figured out where the issue with this approach is.

Could you please elaborate, do you able to offer different approach?
Yes, ```maturin```, for example, requires Rust in order to compile its dependencies.

May be you know the way to download and use pre-build binary packages, for all possible supported OS/architectures?

**Comment by gmabey** [comment]
@amdei This branch worked for me on ubuntu 24.04!
I did have to install some packages that probably already should have been installed: `python3-pkginfo`, `python3-packaging`, `python3-setuptools` and `python3-wheel`.  It looked to me like something was trying to automatically install some of them but that effort (whatever was doing it) produced errors.  However, when I just installed the distro-provided versions, everything worked!

On ubuntu 22.04, ubuntu 20.04, rockylinux 8 and rockylinux 9, I had to both upgrade the distro-provided pip as well as upgrade the above modules before the creation of a .deb from python/pip would work.

So, I conclude that the core of your patch works great, but there may be some dependency auto-foo that is lacking?

Thank you for submitting this MR!

**Comment by amdei** [comment]
Rebased on top of main.
Squashed intermediate commits.

File: lib/fpm/package/python.rb (hunk 5 above)
Comment (amdei) [reply lib/fpm/package/python.rb:321]: It can.
Do you know any cmd option for PEP517 builds to do the same thing as --load-requirements-txt does for setup.py?

File: lib/fpm/package/python.rb (hunk 6 above)
Comment (amdei) [reply lib/fpm/package/python.rb:539]: Seems there is just no alternative for such option (--prefix) in setup.py for PEP517 builds...

File: lib/fpm/package/pyfpm/get_metadata.py (hunk 2 above)
Comment (amdei) [reply lib/fpm/package/pyfpm/get_metadata.py]: Deleted

File: lib/fpm/package/python.rb (hunk 4 above)
Comment (amdei) [reply lib/fpm/package/python.rb]: Deleted

File: lib/fpm/package/python.rb (hunk 3 above)
Comment (amdei) [reply lib/fpm/package/python.rb]: As setup.py project slowly dying, I decide to keep priority of new behavior over old one.

**Comment by amdei** [comment]
> The issue is that all dependencies are being built from source due to a change introduced here: https://github.com/jordansissel/fpm/commit/77e0cf7a768bdd883409d58e0549691a7d1d762f

@cwegener 
Tried to overcome the issue with rebuilding the whole world all over again.

1. Do `pip download` with `--prefer-binary` flag;
2. If wheel was downloaded - do not attempt to rebuild package from sources;
3. If wheel was NOT downloaded - fallback to rebuilding package from sources (either via `setup.py` or `pyproject.toml`).

Can not properly check p. 3 though at the moment...

Could you please try it now?
Build time expected to be _significantly_ lower.
Something like this:
`./bin/fpm  -s python -t deb --python-bin=python3 --python-package-name-prefix python3  openai`

**Comment by deadda7a** [comment]
I would love to have this feature! Sadly I can't contribute due to my lack of Ruby knowledge and time :/

**Comment by jordansissel** [comment]
I had forgotten to check this PR, sorry about that. I think I've done some duplicate work via #2104. For my tests, I *think* #2104/#2105 addresses the same issue that this PR does. I'll include notes for this PR in the changelog for the next release -- thanks to everyone who contributed to both review and coding <3

> Specifically, I will need to test with this PEP-517 project ... https://github.com/openai/openai-python

@cwegener I tested this python project with fpm's main branch (after merging #2104 and #2105) and was able to package it from pypi and from local git checkout and both work:

```
# (at time of writing, using an unreleased fpm version)
% fpm -s python -t deb openai
Created package {:path=>"python3-openai_2.0.1_all.deb"}

# Now try the git repo
% git clone --depth 1 https://github.com/openai/openai-python /tmp/openai-python
% fpm -f -s python -t deb /tmp/openai-python
Created package {:path=>"python3-openai_2.0.1_all.deb"}
```

**Comment by jordansissel** [comment]
Release https://github.com/jordansissel/fpm/releases/tag/v1.17.0 should address this issue. Let me know if you find otherwise :)

**Comment by amdei** [comment]
fpm 1.17.0 failed to convert some packages from Python to deb format.
For example: django 5

It does conversion, but result package is unoperatable.

Going to investigate the issue, and submit new PR.

Hope it will be reviewed this time ;-)

**Comment by jordansissel** [comment]
@amdei I look forward to the details! I tried this myself but it works for me (latest Ubuntu, python3, latest django)

**Comment by amdei** [comment]
@jordansissel spot a source of observed behavior:

If the same package/version already installed on the system, conversion from python to deb silently fails without noticing.
Result package contains only changelog then.

Steps to reproduce (on Debian 12)

1. 
```
# fpm --version
1.17.0
```
2. Install some package (e.g. python-debian) and try to convert it:
```
apt-get update && apt-get install python3-debian
fpm --verbose --log debug --debug --debug-workspace  -s python -t deb python-debian
```
3. Look inside 'converted' package:
```
dpkg-deb -R python-debian_1.0.1_all.deb tmp
ls -la tmp/usr/
```

There is only changelog....


How to fix (not tested (yet)):
pip issue a message:

> Running command {args: ["python", "-m", "pip", "install", "--no-deps", "--root", "/tmp/package-python-staging-268d709e9e6053bf5823da1eb0273fc1bc55d75f6b7d2943d1bdd168dba5", "/tmp/package-python-build-56a7fe6acb3100d867cf610424c6c161ba934ae6ffeecfbac3186b99b0b8/python-debian/python_debian-1.0.1-py3-none-any.whl"], level: :info, file: "fpm/util.rb", line: "139", method: nil}
> Process is running {pid: 12999, level: :debug, file: "fpm/util.rb", line: "148", method: nil}
> Processing /tmp/package-python-build-56a7fe6acb3100d867cf610424c6c161ba934ae6ffeecfbac3186b99b0b8/python-debian/python_debian-1.0.1-py3-none-any.whl {level: :info, file: "cabin/mixins/pipe.rb", line: "47", method: nil}
> **python-debian is already installed with the same version as the provided wheel. Use --force-reinstall to force an installation of the wheel.** {level: :info, file: "cabin/mixins/pipe.rb", line: "47", method: nil}

So, seems `--force-reinstall` should to the trick.

**Comment by amdei** [comment]
> There is only changelog....

Latest main branch do not have such an issue.
Most probably due to introduced usage of `--ignore-installed` pip flag here:
https://github.com/jordansissel/fpm/commit/70377dd177f4509e2187382eacba884371dad005
